File: yami4/src/objc/examples/calculator/server_ns.m
#import <yami4-objc/YAMI4.h>

#import <Foundation/NSAutoreleasePool.h>
#import <Foundation/NSString.h>

#include <stdio.h>


@interface MyHandler : NSObject <YAMI4IncomingMessageCallback>

-(void)call:(YAMI4IncomingMessage *)message;

@end

@implementation MyHandler

-(void)call:(YAMI4IncomingMessage *)message
{
    YAMI4Parameters * params = [message parameters];

    int a = [params getInteger:@"a"];
    int b = [params getInteger:@"b"];

    // prepare the answer with results of four calculations

    YAMI4Parameters * replyParams = [YAMI4Parameters new];

    [replyParams setInteger:@"sum" value:(a + b)];
    [replyParams setInteger:@"difference" value:(a - b)];
    [replyParams setInteger:@"product" value:(a * b)];

    // if the ratio cannot be computed,
    // it is not included in the response
    // the client will interpret that fact properly
    if (b != 0)
    {
        [replyParams setInteger:@"ratio" value:(a / b)];
    }

    [message reply:replyParams];

    printf("got message with parameters %d and %d, "
        "response has been sent back\n", a, b);
}

@end

int main(int argc, char * argv[])
{
    if (argc != 2)
    {
        puts("expecting one parameter: name server address");
        return -1;
    }

    NSAutoreleasePool * pool = [[NSAutoreleasePool alloc] init];
    
    NSString * nameServerAddress =
        [[NSString alloc] initWithUTF8String:argv[1]];

    @try
    {
        YAMI4Agent * serverAgent = [YAMI4Agent new];

        // prepare the server and bind its address
        // to the name server

        NSString * resolvedAddress =
            [serverAgent addListener:@"tcp://*:*"];

        printf("The server is listening on %s\n",
            [resolvedAddress UTF8String]);
            
        YAMI4Parameters * bindParams = [[YAMI4Parameters alloc] init];
        [bindParams setString:@"object" value:@"calculator"];
        [bindParams setString:@"location" value:resolvedAddress];

        YAMI4OutgoingMessage * nsBind =
            [serverAgent send:nameServerAddress
                objectName:@"names" messageName:@"bind"
                parameters:bindParams];

        [nsBind waitForCompletion];
 
        if ([nsBind state] != YAMI4Replied)
        {
            printf("error: %s\n", [[nsBind exceptionMsg] UTF8String]);

            return -1;
        }

        puts("Address bound by name server.");

        MyHandler * myObject = [[MyHandler alloc] init];
        [serverAgent registerObject:@"calculator" callback:myObject];

        // block
        int dummy = getchar();
        (void) dummy;
    }
    @catch (NSException * e)
    {
        printf("error: %s\n", [[e reason] UTF8String]);
    }
    
    [pool drain];
    
    return 0;
}

File: yami4/src/csharp/examples/Calculator/ServerNS/ServerNS.cs
﻿using System;
using System.Threading;
using Inspirel.YAMI;

namespace Calculator
{
    class Server
    {
        private static void calculator(
            object sender, IncomingMessageArgs args)
        {
            // extract the parameters for calculations
            Parameters param = args.Message.Parameters;

            int a = param.GetInteger("a");
            int b = param.GetInteger("b");

            // prepare the answer
            // with results of four calculations
            Parameters replyParams = new Parameters();

            replyParams.SetInteger("sum", a + b);
            replyParams.SetInteger("difference", a - b);
            replyParams.SetInteger("product", a * b);

            // if the ratio cannot be computed,
            // it is not included in the response
            // the client will interpret that fact properly
            if(b != 0)
            {
                replyParams.SetInteger("ratio", a / b);
            }

            args.Message.Reply(replyParams);

            Console.WriteLine(
                "got message with parameters {0} and {1}" +
                ", response has been sent back", a, b);
        }

        static void Main(string[] args)
        {
            if(args.Length != 1)
            {
                System.Console.WriteLine(
                    "expecting one parameter: name server address");
                return;
            }

            string nameServerAddress = args[0];

            try
            {
                Agent serverAgent = new Agent();

                // prepare the server and bind its address
                // to the name server

                String resolvedAddress =
                    serverAgent.AddListener("tcp://*:*");

                System.Console.WriteLine(
                    "The server is listening on {0}",
                    resolvedAddress);

                Parameters bindParams = new Parameters();
                bindParams.SetString("object", "calculator");
                bindParams.SetString("location", resolvedAddress);
                
                OutgoingMessage nsBind = 
                    serverAgent.Send(nameServerAddress,
                            "names", "bind", bindParams);
                
                nsBind.WaitForCompletion();
                if (nsBind.State != 
                    OutgoingMessage.MessageState.REPLIED) {
                    Console.WriteLine("error: {0}" +
                            nsBind.ExceptionMsg);
                    return;
                }
                
                nsBind.Close();

                Console.WriteLine("Address bound by name server.");
                
                serverAgent.RegisterObject("calculator", calculator);

                // block
                while(true)
                {
                    Thread.Sleep(10000);
                }
            }
            catch(Exception ex)
            {
                Console.WriteLine("error: {0}", ex.Message);
            }
        }
    }
}
